theano/tensor/opt_uncanonicalize.py

```diff
@@ -173,3 +173,50 @@ def local_dimshuffle_alloc(node):
 
             return [T.alloc(input_.owner.inputs[0], *new_shape_input)]
     return False
+
+
+@register_uncanonicalize
+@gof.local_optimizer([DimShuffle])
+def local_dimshuffle_subtensor(node):
+    """
+    If a subtensor is inside a dimshuffle which only drop broadcastable dimensions,
+    scrap the dimshuffle and index the subtensor with 0
+
+    x[i:j, :, k:l].dimshuffle(0, 2) => x[i:j, 0, k:l] if x.broadcastable == (False, True, False)
+    """
+    if isinstance(node.op, DimShuffle) and node.inputs[0].owner:
+        # the dimshuffle can only drop dimensions (cannot reshape nor add 'x')
+        if 'x' in node.op.new_order :
+            return False
+        new_order = node.op.new_order
+        past_dim = new_order[0]
```

Comment:
`new_order` may be empty (for instance if all dimensions were broadcastable and we drop them all, to end up with a scalar). In that case, it can be consider "sorted", and the optimization can apply.
You can also try to see if there is a function in Python or NumPy that checks if a list/tuple is sorted.